### PR TITLE
[terminal] permit showing widget for already running terminal

### PR DESCRIPTION
**Pull request by marcdumais-work**
This contains the infrastructure, but no UI yet.

Signed-off-by: Marc Dumais <marc.dumais@ericsson.com>

### CODE REVIEW

**Comment by hexa00** [comment]
The GDB extension will be using that too so we need it in master soon.. 
I'll rework it a bit soon

File: packages/terminal/src/browser/terminal-widget.ts [hunk 1]
```diff
@@ -105,32 +107,23 @@ export class TerminalWidget extends BaseWidget {
...
-            this.title.label = `<terminated>`;
-        };
-        socket.onerror = (err) => {
-            console.error(err);
-        };
-        this.toDispose.push(Disposable.create(() =>
-            socket.close()
-        ));
+        /* An error has occured in the backend.  */
+        if (!success) {
```
Comment (svenefftinge) [inline packages/terminal/src/browser/terminal-widget.ts]: I think await will result in a thrown error if the promise was rejected.

File: packages/terminal/src/browser/terminal-widget.ts [hunk 2]
```diff
@@ -93,6 +93,8 @@ export class TerminalWidget extends BaseWidget {
         (this.term as any).fit()
     }
 
+    // Create a new shell terminal in the back-end and attach it to a
+    // new terminal widget
```
Comment (svenefftinge) [inline packages/terminal/src/browser/terminal-widget.ts]: This should be using jsdoc comment style

File: packages/terminal/src/browser/terminal-widget.ts [hunk 3]
```diff
@@ -105,32 +107,23 @@ export class TerminalWidget extends BaseWidget {
...
-        });
+        this.monitorTerminal(this.terminalId);
+    }
 
-        this.terminalWatcher.onTerminalExit((event: IBaseTerminalExitEvent) => {
-            if (event.terminalId === this.terminalId) {
-                this.title.label = "<terminated>";
-            }
-        });
+    // Attach a terminal widget to an already-running shell terminal
```
Comment (svenefftinge) [inline packages/terminal/src/browser/terminal-widget.ts]: jsdoc / multiline

File: packages/terminal/src/browser/terminal-widget.ts (hunk 1 above)
Comment (hexa00) [reply packages/terminal/src/browser/terminal-widget.ts]: this can't be rejected.

File: packages/terminal/src/browser/terminal-widget.ts [hunk 4]
```diff
@@ -93,45 +93,35 @@ export class TerminalWidget extends BaseWidget {
...
-            this.title.label = `<terminated>`;
-        };
-        socket.onerror = (err) => {
-            console.error(err);
-        };
-        this.toDispose.push(Disposable.create(() =>
-            socket.close()
-        ));
+        this.connectSocket(this.terminalId);
+        this.monitorTerminal(this.terminalId);
```
Comment (svenefftinge) [inline packages/terminal/src/browser/terminal-widget.ts]: Is there a reason you reversed the order? Don't we maybe miss an event?

Comment (hexa00) [reply packages/terminal/src/browser/terminal-widget.ts]: The order doesn't matter, both event sources are different.

Comment (svenefftinge) [reply packages/terminal/src/browser/terminal-widget.ts]: ok

**Comment by hexa00** [comment]
@marcdumais-work this worked for me with the gdb extension, could you confirm that it still works for you for tasks ?

And if we so we can merge it ?

**Comment by marcdumais-work** [comment]
@hexa00 I just tests with Tasks, and it worked. You can merge. Thanks!

**Pull request merged**